E2E framework cleanups

thockin opened the pull request:
A few things have always bugged me about our framework.  Specifically the guarantees about cleanup are almost useless in the face of test timeouts or manual ^C (I assume a timeout == SIGINT, but I guess we'll see?).

Now this tries harder to always clean up. 

Fixes #14944

thockin committed:
e2e: Call AfterEach handlers in case of an abort

I was tired of aborted tests leaving debris in my cluster.

thockin committed:
Move "extra" namespaces into framework

Now an aborted test will delete namespaces.  Hopefully this means less GCE
leaks.

thockin commented:
@kubernetes/sig-testing @ihmccreery @ixdy @quinton-hoole

thockin commented:
I also think we can make more tests simpler and safer by internalizing stuff into Framework.

ihmccreery commented on test/e2e/e2e.go:
```diff
@@ -188,14 +189,48 @@ var _ = ginkgo.SynchronizedBeforeSuite(func() []byte {
+var lock sync.Mutex
+var cleanupActions = map[*int]func(){}
+
+func AddCleanupAction(fn func()) *int {
```
Should have comment explanation.

ihmccreery commented on test/e2e/e2e.go:
```diff
@@ -188,14 +189,48 @@ var _ = ginkgo.SynchronizedBeforeSuite(func() []byte {
+	return p
+}
+
+func RemoveCleanupAction(p *int) {
```
Should have comment explanation.

ihmccreery commented on test/e2e/e2e.go:
```diff
@@ -188,14 +189,48 @@ var _ = ginkgo.SynchronizedBeforeSuite(func() []byte {
+	delete(cleanupActions, p)
+}
+
+func RunCleanupActions() {
```
Should have comment explanation.

ihmccreery commented on test/e2e/e2e.go:
```diff
@@ -188,14 +189,48 @@ var _ = ginkgo.SynchronizedBeforeSuite(func() []byte {
+var cleanupActions = map[*int]func(){}
+
+func AddCleanupAction(fn func()) *int {
+	p := new(int)
```
Is there really no better way to get a set object (which is what, as far as I can tell, you're doing).

thockin replied:
I want a handle that is guaranteed unique - pointers seems to be cheap and easy.  I'll put it behind a type.

ihmccreery commented on test/e2e/example_cluster_dns.go:
```diff
@@ -75,15 +75,8 @@ var _ = Describe("ClusterDns [Feature:Example]", func() {
-			} else {
-				Logf("Found DeleteNamespace=false, skipping namespace deletion!")
-			}
+			namespaces[i], err = framework.CreateNamespace(fmt.Sprintf("dnsexample%d", i), c, nil)
```
The comment on line 74 seems like a smell to me, along with framework.go:43.  Why not have Framework just have a slice to begin with, and you can tack on additional namespaces as necessary?

thockin replied:
I didn't want to go through and audit all the tests - I just wanted to make the smallest and most obvious change I could.  In this case, there's some value to the names, maybe?

ihmccreery commented on test/e2e/framework.go:
```diff
@@ -200,6 +214,14 @@ func (f *Framework) afterEach() {
 }
 
+func (f *Framework) CreateNamespace(baseName string, c *client.Client, labels map[string]string) (*api.Namespace, error) {
+	ns, err := createTestingNS(baseName, c, labels)
```
Seems like `createTestingNS` should just be pulled into `framework.go` and de-duped with this function.

ixdy replied:
+1 for just deleting `createTestingNS`.

thockin replied:
It is used in a few places NOT using Framework.  There's value to having layers of behavior, i think.

ixdy replied:
The one remaining use of `createTestingNS` is `test/e2e/namespace.go`, which *should* be using Framework. (Everything should be using Framework.) But we can punt that, I guess. sigh.

ixdy replied:
Cleaning up all tests to use `Framework` in #20939. `namespace.go` will require a little bit more refactoring to have it use `f.CreateNamespace`.

thockin replied:
can this PR proceed until then?

ixdy replied:
`namespace.go` is now using `Framework`. I'd still argue that we should update it to use this function instead of `createTestingNS`, though I leave it up to you whether to punt that to another PR or do here.

ihmccreery commented:
Mostly LGTM.  Public functions should have comments supporting them.  And I think factoring `createTestingNS` into `framework.go` would help things.

ihmccreery commented:
cc @ixdy since he's much more familiar with this code than I am.

ixdy commented on test/e2e/e2e.go:
```diff
@@ -188,14 +189,48 @@ var _ = ginkgo.SynchronizedBeforeSuite(func() []byte {
 }, func() {
 	// Run only Ginkgo on node 1
+
+	RunCleanupActions()
```
this will need to run on all nodes, since they are actually separate processes. (move to the other block.)

thockin replied:
I don't know what a node is here, but I trust you :)

ixdy replied:
Ginkgo runs tests in parallel by spawning several `ginkgo` processes, or nodes, which is what this is referring to.

Maybe they should've been called minions. :)

ixdy commented on test/e2e/e2e.go:
```diff
@@ -188,14 +189,48 @@ var _ = ginkgo.SynchronizedBeforeSuite(func() []byte {
 
 })
 
+var lock sync.Mutex
```
`lock` seems like a very generic name, given how much we have in the `e2e` package.

thockin replied:
ok

ixdy commented:
It feels like this is a bug in Ginkgo that should really be fixed there, but I don't know if we're "doing something wrong" which is why it doesn't work this way yet.

thockin commented:
re being a bug in ginkgo, I agree, but I could not find out who to fix it, and I think we want it fixed.  So commented.

thockin commented:
new push is up with comments fixed or rebutted (I hope :)

ixdy commented:
re: Ginkgo - we can open an issue over at https://github.com/onsi/ginkgo/issues and do this workaround at the same time; @onsi is pretty responsive.

thockin commented:
Bug filed

ihmccreery commented on test/e2e/e2e.go:
```diff
@@ -189,12 +190,54 @@ var _ = ginkgo.SynchronizedBeforeSuite(func() []byte {
+// whole test being terminated.  This allows arbitrary pieces of the overall
+// test to hook into SynchronizedAfterSuite().
+func AddCleanupAction(fn func()) CleanupActionHandle {
+	p := new(int)
```
Is there a way to type `p` better here?  It'd be better to use `CleanupActionHandle`.

thockin replied:
I could make the handle be int and return a pointer to handle, but that's
not right.  I added a cast (frankly I am not sure why it works without the
cast).

On Tue, Feb 9, 2016 at 9:34 AM, Isaac Hollander McCreery <
notifications@github.com> wrote:

> In test/e2e/e2e.go
> <https://github.com/kubernetes/kubernetes/pull/20771#discussion_r52343173>
> :
>
> > @@ -189,12 +190,54 @@ var _ = ginkgo.SynchronizedBeforeSuite(func() []byte {
> >
> >  })
> >
> > +type CleanupActionHandle *int
> > +
> > +var cleanupActionsLock sync.Mutex
> > +var cleanupActions = map[CleanupActionHandle]func(){}
> > +
> > +// AddCleanupAction installs a function that will be called in the event of the
> > +// whole test being terminated.  This allows arbitrary pieces of the overall
> > +// test to hook into SynchronizedAfterSuite().
> > +func AddCleanupAction(fn func()) CleanupActionHandle {
> > +	p := new(int)
>
> Is there a way to type p better here? It'd be better to use
> CleanupActionHandle.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/20771/files#r52343173>.
>

ihmccreery commented on test/e2e/framework.go:
```diff
@@ -74,6 +81,9 @@ func NewFramework(baseName string) *Framework {
 
 // beforeEach gets a client and makes a namespace.
 func (f *Framework) beforeEach() {
+	// The fact that we need this feels like a bug in ginkgo.
```
xref issue filed against Ginkgo.

ixdy replied:
can you update this comment with the xref?

thockin replied:
https://github.com/onsi/ginkgo/issues/222 - added to next push

ihmccreery commented:
LGTM minus nits; I'll leave to @ixdy to do final review.

thockin commented:
rebased.  Looking for LGTM

thockin commented:
I have fixed comments and rebased.  I am out of town next week, so I will not have a chance to rebase again.  I might just manually merge once this passes tests.
